sphinx/domains/index.py

```diff
@@ -68,12 +69,19 @@ class IndexDirective(SphinxDirective):
     required_arguments = 1
     optional_arguments = 0
     final_argument_whitespace = True
-    option_spec = {}  # type: Dict
+    option_spec = {
+        'name': directives.unchanged,
+    }  # type: Dict
 
     def run(self) -> List[Node]:
         arguments = self.arguments[0].split('\n')
         targetid = 'index-%s' % self.env.new_serialno('index')
         targetnode = nodes.target('', '', ids=[targetid])
+
+        if 'name' in self.options:
+            targetname = self.options['name']
+            targetnode['names'].append(targetname)
```

Comment:
@shimizukawa Do you have any opinion for this?

We can also give a node id based on given name like this:
```
if 'name' in self.options:
    targetname = self.options['name']
    targetnode = nodes.target('', '', names=[targetname])
else:
    targetid = 'index-%s' % self.env.new_serialno('index')
    targetnode = nodes.target('', '', ids=[targetid])
```

pros:
* We can use natural node_id generated from name. Therefore HTML anchor would be also natural.

cons:
* ID generation rule will be changed. As a result, some of URLs for indices will be changed.

Reply:
For cons, `index-<serialno>` name will be changed by many reasons (new index entry is inserted before existing index entries)  so `index-<serialno>` is not a PERMANENT.

So, +1 for `:name:`  option to get PERMANENT target.

Reply:
Okay, go with above my code.

Reply:
@geographika Could you update this please?

Reply:
Done - although I used targetid rather than targetname as targetid is required later and I would get a `UnboundLocalError` at `indexnode['entries'].extend(process_index_entry(entry, targetid))`

Reply:
Please use `targetnode['ids'][0]' instead. `document.note_implicit_target()` generates node_id automatically if not assigned yet. So we can use it after doing it.